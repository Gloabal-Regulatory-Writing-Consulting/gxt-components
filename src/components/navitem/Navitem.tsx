import { FC } from "react";
import {
  NavIcon,
  NavItemContainer,
  NavLinkStyled,
  NavText,
} from "./NavItemStyledComponents";

export interface INavItem {
  navigateTo: string;
  text: string;
  Icon?: React.ComponentType<any>;
  className?: string;
  onClickHandler?: () => void;
}

export interface NavItemProps extends INavItem {
  isExpanded: boolean;
  isLinkActive: (path: string) => boolean;
}

export interface IClientStyling {
  thumbnail: string;
  logo: string;
}

const Navitem: FC<NavItemProps> = ({
  navigateTo,
  text,
  Icon,
  isExpanded,
  isLinkActive,
  onClickHandler,
}) => {
  const isLinkActivated = isLinkActive(navigateTo);

  return (
    <div onClick={onClickHandler} data-testid={text + "-navlink"}>
      <NavLinkStyled
        data-tooltip-id="navlink"
        data-tooltip-content={text}
        to={navigateTo}
        data-testid={text}
        end
      >
        <NavItemContainer $isLinkActive={isLinkActivated}>
          <NavIcon>
            {Icon && (
              <Icon
                stroke={
                  isLinkActivated ? "var(--primary-200, #177BA6)" : undefined
                }
                fill={
                  isLinkActivated ? "var(--primary-200, #177BA6)" : undefined
                }
              ></Icon>
            )}
          </NavIcon>
          <NavText $isExpanded={isExpanded} $isActive={isLinkActivated}>
            {text}
          </NavText>
        </NavItemContainer>
      </NavLinkStyled>
    </div>
  );
};

export default Navitem;
