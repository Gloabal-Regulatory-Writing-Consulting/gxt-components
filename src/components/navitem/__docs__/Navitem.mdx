import { Canvas, Meta, Controls } from "@storybook/blocks";
import * as Navitem from "./Navitem.stories.tsx";

<Meta of={Navitem} title="Navitem" />

# Navitem

## Example

<Canvas of={Navitem.NavitemComponent} />

## Usage

```tsx
import Navitem from "sld-ui";
import { ReactComponent as HomeIcon } from "./icons/HomeIcon.svg";

const Example = () => {
  const isLinkActive = (path) => {
    // return true if path matches the current path of application
  };
  return (
    <Navitem
      navigateTo="/home"
      text="Home"
      Icon={HomeIcon}
      isExpanded={true}
      isLinkActive={isLinkActive}
    />
  );
};

export default Example;
```

## Props
Navitem component with different props.
<Controls of={Navitem.NavitemComponent} />
