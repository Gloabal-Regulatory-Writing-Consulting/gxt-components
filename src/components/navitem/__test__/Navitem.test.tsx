import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import Navitem, { NavItemProps } from "../Navitem";
import { describe, expect, it, vi } from "vitest";

const MockIcon = (props: any) => <svg {...props} data-testid="mock-icon" />;

const mockOnClickHandler = vi.fn();

const renderWithRouter = (ui: React.ReactElement, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);

  return render(
    <MemoryRouter initialEntries={[route]}>
      <Routes>
        <Route path="*" element={ui} />
      </Routes>
    </MemoryRouter>,
  );
};

const defaultProps: NavItemProps = {
  navigateTo: "/home",
  text: "Home",
  Icon: MockIcon,
  className: "",
  isExpanded: true,
  isLinkActive: (path: string) => path === "/home",
  onClickHandler: mockOnClickHandler,
};

describe("Navitem component", () => {
  it("renders without crashing", () => {
    const { asFragment } = renderWithRouter(<Navitem {...defaultProps} />);
    expect(screen.getByTestId("Home")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with correct text", () => {
    const { asFragment } = renderWithRouter(<Navitem {...defaultProps} />);
    expect(screen.getByText("Home")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders icon when provided", () => {
    const { asFragment } = renderWithRouter(<Navitem {...defaultProps} />);
    expect(screen.getByTestId("mock-icon")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not render icon when not provided", () => {
    const { asFragment } = renderWithRouter(
      <Navitem {...defaultProps} Icon={undefined} />,
    );
    expect(screen.queryByTestId("mock-icon")).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders expanded text when isExpanded is true", () => {
    const { asFragment } = renderWithRouter(<Navitem {...defaultProps} />);
    expect(screen.getByText("Home")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not render expanded text when isExpanded is false", async () => {
    const { asFragment } = renderWithRouter(
      <Navitem {...defaultProps} isExpanded={false} />,
    );
    const navText = screen.getByText("Home");
    expect(navText).toHaveStyle("visibility: hidden");
    expect(asFragment()).toMatchSnapshot();
  });

  it("applies active styles when link is active", () => {
    const { asFragment } = renderWithRouter(<Navitem {...defaultProps} />, {
      route: "/home",
    });
    const navItemContainer = screen.getByTestId("Home").firstChild;
    expect(navItemContainer).toHaveStyle("border-left: solid");
    expect(navItemContainer).toHaveStyle(
      "border-left-color: var(--primary-200, #177BA6)",
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not apply active styles when link is not active", () => {
    const isLinkActiveMock = (path: string) => path !== "/home";
    const { asFragment } = renderWithRouter(
      <Navitem {...defaultProps} isLinkActive={isLinkActiveMock} />,
      {
        route: "/another-route",
      },
    );
    const navItemContainer = screen.getByTestId("Home").firstChild;
    expect(navItemContainer).not.toHaveStyle("border-left: solid");
    expect(asFragment()).toMatchSnapshot();
  });

  it("applies correct icon color when link is active", () => {
    const { asFragment } = renderWithRouter(<Navitem {...defaultProps} />, {
      route: "/home",
    });
    const icon = screen.getByTestId("mock-icon");
    expect(icon).toHaveAttribute("stroke", "var(--primary-200, #177BA6)");
    expect(icon).toHaveAttribute("fill", "var(--primary-200, #177BA6)");
    expect(asFragment()).toMatchSnapshot();
  });

  it("applies correct icon color when link is not active", () => {
    const isLinkActiveMock = (path: string) => path !== "/home";
    const { asFragment } = renderWithRouter(
      <Navitem {...defaultProps} isLinkActive={isLinkActiveMock} />,
      {
        route: "/another-route",
      },
    );
    const icon = screen.getByTestId("mock-icon");
    expect(icon).not.toHaveAttribute("stroke", "var(--primary-200, #177BA6)");
    expect(icon).not.toHaveAttribute("fill", "var(--primary-200, #177BA6)");
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the onClickHandler when clicked", () => {
    renderWithRouter(<Navitem {...defaultProps} />);
    fireEvent.click(screen.getByTestId("Home-navlink"));
    expect(mockOnClickHandler).toHaveBeenCalledTimes(1);
  });
});
