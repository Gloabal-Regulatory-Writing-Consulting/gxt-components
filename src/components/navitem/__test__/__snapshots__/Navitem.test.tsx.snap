// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navitem component > applies active styles when link is active 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      aria-current="page"
      class="sc-gEvDqW ixgtyG active"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > applies correct icon color when link is active 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      aria-current="page"
      class="sc-gEvDqW ixgtyG active"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > applies correct icon color when link is not active 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cExATB"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
          />
        </div>
        <p
          class="sc-fqkwJk fFtWrx"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > does not apply active styles when link is not active 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cExATB"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
          />
        </div>
        <p
          class="sc-fqkwJk fFtWrx"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > does not render expanded text when isExpanded is false 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk fDjqbC"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > does not render icon when not provided 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        />
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > renders expanded text when isExpanded is true 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > renders icon when provided 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > renders with correct text 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Navitem component > renders without crashing 1`] = `
<DocumentFragment>
  <div
    data-testid="Home-navlink"
  >
    <a
      class="sc-gEvDqW ixgtyG"
      data-testid="Home"
      data-tooltip-content="Home"
      data-tooltip-id="navlink"
      href="/home"
    >
      <div
        class="sc-aYaIB cCERKD"
      >
        <div
          class="sc-eqUzNf kbmWsj"
        >
          <svg
            data-testid="mock-icon"
            fill="var(--primary-200, #177BA6)"
            stroke="var(--primary-200, #177BA6)"
          />
        </div>
        <p
          class="sc-fqkwJk lmZzif"
        >
          Home
        </p>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
